modules/highlight/highlight.go

```diff
@@ -196,18 +196,25 @@ func File(numLines int, fileName, language string, code []byte) []string {
 	}
 
 	m := make([]string, 0, numLines)
-	for _, v := range strings.SplitN(htmlbuf.String(), "\n", numLines) {
+	for i, v := range strings.SplitN(htmlbuf.String(), "\n", numLines) {
 		content := string(v)
+
+		// remove useless wrapper nodes that are always present
+		content = strings.Replace(content, "<span class=\"line\"><span class=\"cl\">", "", 1)
+		content = strings.TrimPrefix(content, `</span></span>`)
+
+		// if there's no final newline, closing tags will be on last line
+		if !finalNewLine && i == numLines-1 {
+			content = strings.TrimSuffix(content, `</span></span>`)
+		}
```

Comment:
These two trimmings of `</span></span>` seems to be fragile. Maybe these tags could be removed more accurately

Reply:
I guess only with an XML parser it could be made more robust, but also likely much slower. Or maybe Chroma has an option to emit less wrapper tags, I did not investigate.

Reply:
> I guess only with an XML parser it could be made cleaner

FYI: https://pkg.go.dev/golang.org/x/net/html

Reply:
I have looked at the Chroma code, it seems impossible to get rid of the line wrapper tags (two `span`s)  at the moment.

Since the outputted HTML is strict, I think it can be processed accurately by a simple `<tag>` match and a stack.

Actually, the question in my mind is that why the Replace+TrimPrefix+TrimSuffix should be used. When reading the code, I can not guess the correctness for it. Even a `TrimPrefix("<span ...")` + `TrimPrefix("</span>")` is more readable and can be more stable.

Reply:
The final TrimSuffix is to prevent unbalanced HTML from being emitted, which I think the previous code did. Browsers generally ignore unmatched closing tags, that's probably why it was never noticed.

What could maybe be done is instead of splitting and iterating these "lines", regexp replacement across those whole string could be applied. It might even be a bit faster overall.

Ultimately, I consider the lack of newline characters on empty line a Chroma bug.

Reply:
I am asking for the definition because I have some ideas to rewrite the highlight code, to make everything clear, instead of patching the buggy code again and again.

I would propose a PR for it, including the new line problem and the empty line problem.